x/cronos/keeper/precompiles/ica.go

```diff
@@ -0,0 +1,199 @@
+package precompiles
+
+import (
+	"bytes"
+	"errors"
+	"math/big"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	icacontrollerkeeper "github.com/cosmos/ibc-go/v7/modules/apps/27-interchain-accounts/controller/keeper"
+	icacontrollertypes "github.com/cosmos/ibc-go/v7/modules/apps/27-interchain-accounts/controller/types"
+	icatypes "github.com/cosmos/ibc-go/v7/modules/apps/27-interchain-accounts/types"
+	ibcchannelkeeper "github.com/cosmos/ibc-go/v7/modules/core/04-channel/keeper"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/vm"
+	"github.com/evmos/ethermint/x/evm/keeper/precompiles"
+)
+
+var (
+	RegisterAccountMethod abi.Method
+	QueryAccountMethod    abi.Method
+	SubmitMsgsMethod      abi.Method
+	IcaContractAddress    = common.BytesToAddress([]byte{102})
+)
+
+func init() {
+	addressType, _ := abi.NewType("address", "", nil)
+	stringType, _ := abi.NewType("string", "", nil)
+	bytesType, _ := abi.NewType("bytes", "", nil)
+	uint256Type, _ := abi.NewType("uint256", "", nil)
+	RegisterAccountMethod = abi.NewMethod(
+		"registerAccount", "registerAccount", abi.Function, "", false, false, abi.Arguments{abi.Argument{
+			Name: "connectionID",
+			Type: stringType,
+		}, abi.Argument{
+			Name: "owner",
+			Type: addressType,
+		}, abi.Argument{
+			Name: "version",
+			Type: stringType,
+		}},
+		abi.Arguments{abi.Argument{
+			Name: "res",
+			Type: bytesType,
+		}},
+	)
+	QueryAccountMethod = abi.NewMethod(
+		"queryAccount", "queryAccount", abi.Function, "", false, false, abi.Arguments{abi.Argument{
+			Name: "connectionID",
+			Type: stringType,
+		}, abi.Argument{
+			Name: "owner",
+			Type: addressType,
+		}},
+		abi.Arguments{abi.Argument{
+			Name: "res",
+			Type: bytesType,
+		}},
+	)
+	SubmitMsgsMethod = abi.NewMethod(
+		"submitMsgs", "submitMsgs", abi.Function, "", false, false, abi.Arguments{abi.Argument{
+			Name: "connectionID",
+			Type: stringType,
+		}, abi.Argument{
+			Name: "owner",
+			Type: addressType,
+		}, abi.Argument{
+			Name: "data",
+			Type: stringType,
+		}, abi.Argument{
+			Name: "timeout",
+			Type: uint256Type,
+		}},
+		abi.Arguments{abi.Argument{
+			Name: "res",
+			Type: bytesType,
+		}},
+	)
+}
+
+type IcaContract struct {
+	cdc                 codec.Codec
+	channelKeeper       *ibcchannelkeeper.Keeper
+	icaControllerKeeper *icacontrollerkeeper.Keeper
+}
+
+func NewIcaContract(
+	cdc codec.Codec,
+	channelKeeper *ibcchannelkeeper.Keeper,
+	icaControllerKeeper *icacontrollerkeeper.Keeper,
+) precompiles.StatefulPrecompiledContract {
+	return &IcaContract{
+		cdc,
+		channelKeeper,
+		icaControllerKeeper,
+	}
+}
+
+func (ic *IcaContract) Address() common.Address {
+	return IcaContractAddress
+}
+
+// RequiredGas calculates the contract gas use
+func (ic *IcaContract) RequiredGas(input []byte) uint64 {
+	return RelayerContractRequiredGas
```

Comment:
is this TODO in this PR?

Reply:
yes for now, can use another const for ica

Reply:
maybe add a commit for this line for explaining why uses a const instead of calculating the input byte array.